### PR TITLE
Fa 17/account rate limiting

### PR DESCRIPTION
**Pull request by dempe**
@simplymeasured/team-narwhal  review please.

~¡Still WIP!:~
- [x] Update sm-user-rate-limiting postgres increment function
- [x] Remove API ID from sm-user-rate-limiting
- [x] Separate client_id and account_id in the sm-account-rate-limiting DB schema
- [x] Rename sm-rate-limiting to sm-user-rate-limiting

Tested locally.

This includes the rate limiting updates for both _user_ and _account_ rate limits.

- For *user* rate limits (i.e., "the Viper rate limit"), this is assuming that we will be using only the "local" policy.  The local policy is the most performant of their three policies (local, redis, cluster), but it is the least accurate.  It is essentially for protecting back-end services which is precisely why we created this rate limit in the first place.
- For the *account* rate limits (i.e., "the API access rate limits"), this is assuming that we will only be using the "cluster" policy.  The cluster policy is the most accurate of their three policies, but it is the least performant.  Compared to the Viper rate limiting, this rate limit has fairly low overhead, and, since we are charging customers for increased rate limits, we need higher accuracy here, so we fit the "cluster" use case quite well.

For more information on Kong rate limits, their policies, and their use cases, see:  https://getkong.org/plugins/rate-limiting/

Note that we still need to do some perf testing on this.  That work is encompassed in a forthcoming ticket.  I have removed the Redis logic, but it can be added back in quite easily if needed.

Also, all references to Cassandra have been removed for clarity.

### CODE REVIEW

**Comment by dempe** [comment]
Reminder, incorporate feedback from:  https://github.com/simplymeasured/kong-plugin/pull/1

**Comment by dempe** [comment]
Another suggestion:  since we have `sm-account-rate-limiting`, this would be a good time to rename `sm-rate-limiting` to `sm-user-rate-limiting` to be more descriptive.

**Comment by dempe** [comment]
@ThorArakis Looking at the diffs now.
* In `get_identifier(conf)`, they are passing in `conf`, because they have a param, `limit_by` in the schema that they're using.  Since we already know what to limit by, we do not need to pass this in.
* In `get_usage(conf, api_id, identifier, current_timestamp, limits)`, they are passing `api_id`, because they use that as part of their primary key in the rate limit table, but I think we want our rate limits to apply across APIs. Likewise for `increment`, et al.
* There should be no functional difference in the log statements - they added an alias, `local ngx_log = ngx.log`, on line 10.  Not sure how `ngx_log` is any easier, but  ¯\_(ツ)_/¯

File: kong/plugins/sm-account-rate-limiting/handler.lua [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    RateLimitingHandler.super.access(self)
+    local current_timestamp = timestamp.get_utc()
+
+    -- NOTE:  the order in which the following fields are parsed is important, as account_id will not be present in UI JWTs
+    -- If the client_id in the JWT does not match the client ID specified in the plugin config, we will continue.
+    -- This allows us to limit only certain clients (e.g., developer tokens)
+    local jwt = get_jwt()
+    local client_id = parse_claim(jwt, "aud")
+
+    if not (client_id == conf.client_id) then
```
Comment (ThorArakis) [inline kong/plugins/sm-account-rate-limiting/handler.lua]: Do we still need to do this if we can associate a rate-limit plugin with a specific `consumer_id` already natively?
https://github.com/simplymeasured/kong-plugin/pull/2/files#diff-45f543026950b933b0a27eae9bf2e986R51

File: kong/plugins/sm-account-rate-limiting/handler.lua [hunk 2]
```diff
@@ -0,0 +1,150 @@
+-- Copyright (C) Mashape, Inc.
+
+local BasePlugin = require "kong.plugins.base_plugin"
+local policies = require "kong.plugins.sm-account-rate-limiting.policies.init"
```
Comment (dempe) [inline kong/plugins/sm-account-rate-limiting/handler.lua]: I had it that way initially, but Kong kept trying to access `kong.plugins.sm-account-rate-limiting.policies.lua`, which doesn't exist.  The other file in the `policies` directory is `cluster`, but that is only referenced from the `init.lua` file.  So everything should be directly or indirectly accessible this way.
Not actually sure how the Mashape version is working by just referencing the directory... I could not figure out how to get that to work.

File: kong/plugins/sm-user-rate-limiting/daos.lua [hunk 3]
```diff
@@ -0,0 +1,3 @@
+return {
+  tables = {"sm_user_ratelimiting_metrics"}
```
Comment (dempe) [inline kong/plugins/sm-user-rate-limiting/daos.lua]: Hmm.  Good catch.

File: kong/plugins/sm-account-rate-limiting/handler.lua (hunk 1 above)
Comment (dempe) [inline kong/plugins/sm-account-rate-limiting/handler.lua]: Good point.

Comment (dempe) [inline kong/plugins/sm-account-rate-limiting/handler.lua]: Err. I already removed the `limit_by` code.  I think this way is simpler.

**Pull request merged**